report/report.go

```diff
@@ -340,7 +345,11 @@ func (r Report) Validate() error {
 // Upgrade returns a new report based on a report received from the old probe.
 //
 // This for now creates node's LatestControls from Controls.
-func (r Report) Upgrade() Report {
+func (r Report) Upgrade(appVersion string) Report {
+	if !r.RequiresUpgrade(appVersion) {
```

Comment:
I reckon `RequiresUpgrade` is too high-level an API. We will eventually have multiple different kinds of upgrade to perform, depending on the version. E.g. consider how you'd expand this function to deal with #2960.

I reckon a `err, semver := Report.ScopeVersion()` function that returns the lowest scope semver or error (if none can be found) would work better.

Then this function can dispatch different upgrade tasks depending on comparing that version to specific known versions like `latestControlsRelease`.

Reply:
There is another problem with the approach here. In #2959 I suggested that upgrades should take place in the "ingest" path, so we store upgraded reports. But the check here would not notice that, so upgrades would still be performed when an already-upgraded stored report is subsequently read.